module/advancement/types/scale-value.mjs

```diff
@@ -2,6 +2,195 @@ import Advancement from "../advancement.mjs";
 import AdvancementConfig from "../advancement-config.mjs";
 import AdvancementFlow from "../advancement-flow.mjs";
 
+/**
+ * Class responsible for formatting and presenting a scale value of a specific type.
+ * @param {ScaleValueAdvancement} advancement  The advancement to which this scale value belongs.
+ * @param {object} data                        Data for a scale value at a specific level.
+ */
+export class ScaleValueType {
+  constructor(advancement, data) {
+    Object.assign(this, data);
+    this.advancement = advancement;
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Information on how a scale value of this type is configured.
+   *
+   * @typedef {object} ScaleValueTypeMetadata
+   * @property {string} label       Name of this type.
+   * @property {string} hint        Hint for this type shown in the scale value configuration.
+   * @property {boolean} isNumeric  When using the default editing interface, should numeric inputs be used?
+   */
+
+  /**
+   * Configuration information for this scale value type.
+   * @type {ScaleValueTypeMetadata}
+   */
+  static get metadata() {
+    return {
+      label: "DND5E.AdvancementScaleValueTypeString",
+      hint: "DND5E.AdvancementScaleValueTypeHintString",
+      isNumeric: false
+    };
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Attempt to convert another scale value type to this one.
+   * @param {ScaleValueType} original  Original type to attempt to convert.
+   * @returns {object|null}
+   */
+  static converted(original) {
+    return { value: original.prepared };
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * This scale value prepared to be used in roll formulas.
+   * @type {string|null}
+   */
+  get prepared() { return this.value; }
+
+  /* -------------------------------------------- */
+
+  /**
+   * This scale value formatted for display.
+   * @type {string|null}
+   */
+  get formatted() { return this.prepared; }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Shortcut to the prepared value when used in roll formulas.
+   * @returns {string}
+   */
+  toString() {
+    return this.prepared;
+  }
+}
+
+
+export class ScaleValueTypeNumber extends ScaleValueType {
+
+  static get metadata() {
+    return foundry.utils.mergeObject(super.metadata, {
+      label: "DND5E.AdvancementScaleValueTypeNumber",
+      hint: "DND5E.AdvancementScaleValueTypeHintNumber",
+      isNumeric: true
+    });
+  }
+
+  /* -------------------------------------------- */
+
+  static converted(original) {
+    const value = Number(original.prepared);
+    if ( value.isNaN ) return null;
```

Comment:
Is this really a property on `Number` instances? I think this should be `isNaN(value)` or `Number.isNaN(value)` (I think they're functionally identical given that we've already cast to `Number` here).